File: py-algorithm/number/simplified_fractions.py
import unittest
from typing import List
from fractions import Fraction
from math import gcd

"""
    https://leetcode.com/problems/simplified-fractions/
    
    Cho số nguyên n. Tìm tất cả các phân số thuộc (0,1) có mẫu số <= n 
"""


class Solution(unittest.TestCase):

    @staticmethod
    def simplifiedFractions(n: int) -> List[str]:
        result = set()
        for i in range(1, n):
            for j in range(i + 1, n + 1):
                if gcd(i, j) == 1:
                    f = f"{i}/{j}"
                    result.add(f)

        return list(result)

    def test(self):
        self.assertEqual(["1/2"], self.simplifiedFractions(2))
        # self.assertEqual(["1/2", "1/3", "2/3"], self.simplifiedFractions(3))
        # self.assertEqual(["1/2", "1/3", "1/4", "2/3", "3/4"], self.simplifiedFractions(4))


if __name__ == '__main__':
    unittest.main()

File: leetcode/src/main/java/number/SimplifiedFractions.java
package number;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

/**
 * https://leetcode.com/problems/simplified-fractions/
 *
 * Cho số nguyên dương n. Tìm tất cả phân số thuộc (0, n) có mẫu số <= n.
 */
public class SimplifiedFractions {

    // a = b * x + y -> gcd(a,b) = gcd(b, a % b)
    public int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    public List<String> simplifiedFractions(int n) {
        Set<String> result = new HashSet<>();
        for (int i = 1; i <= n - 1; i++) { // Chỉ cần chạy đến n - 1
            for (int j = i + 1; j <= n; j++) { // Chạy đến n
                // int gcd = gcd(i, j);
                if (gcd(i, j) == 1) {
                    // String f = i + "/" + j;
                    result.add(i + "/" + j);
                }
            }
        }
        return new ArrayList<>(result);
    }

    @Test
    public void test() {
        assertIterableEquals(List.of("1/2"), simplifiedFractions(2));
    }

}
